File: Atividades A1/15 Questões/Java/Questoes15/src/Questao7/Main.java
package Questao7;
//7. Associação Crie uma classe Escola e uma classe Professor. A associação deve permitir que uma escola tenha vários professores, e um professor possa lecionar em várias escolas.

import java.util.ArrayList;
import java.util.List;

class Professor {
    private String nome;
    private List<Escola> escolas;

    public Professor(String nome) {
        this.nome = nome;
        this.escolas = new ArrayList<>();
    }

    public void addEscola(Escola escola) {
        if (!this.escolas.contains(escola)) {
            this.escolas.add(escola);
            escola.addProfessor(this);
        }
    }

    public String getNome() {
        return nome;
    }

    public List<Escola> getEscolas() {
        return escolas;
    }
}

class Escola {
    private String nome;
    private List<Professor> prof;

    public Escola(String nome) {
        this.nome = nome;
        this.prof = new ArrayList<>();
    }

    public String getNome() {
        return nome;
    }

    public List<Professor> getProf() {
        return prof;
    }

    public void addProfessor(Professor professor) {
        if (!this.prof.contains(professor)) {
            this.prof.add(professor);
            professor.addEscola(this);
        }
    }
}

// Cliente
public class Main {
    public static void main(String[] args) {

        Professor prof1 = new Professor("Kate");
        Professor prof2 = new Professor("Jack");
        Professor prof3 = new Professor("John");
        Escola escola1 = new Escola("Escola Ilha");
        Escola escola2 = new Escola("Escola Dharma");

        prof1.addEscola(escola1);
        prof2.addEscola(escola1);
        prof2.addEscola(escola2);
        prof3.addEscola(escola2);

        System.out.println("Escolas de Kate:");
        for (Escola escola : prof1.getEscolas()) System.out.println(escola.getNome());
        System.out.println("\nEscolas de Jack:");
        for (Escola escola : prof2.getEscolas()) System.out.println(escola.getNome());
        System.out.println("\nEscolas de Jhon:");
        for (Escola escola : prof3.getEscolas()) System.out.println(escola.getNome());

        System.out.println("\nProfessores na " + escola1.getNome() + ":");
        for (Professor professor : escola1.getProf()) System.out.println(professor.getNome());
        System.out.println("\nProfessores na " + escola2.getNome() + ":");
        for (Professor professor : escola2.getProf()) System.out.println(professor.getNome());

    }
}

File: Atividades A1/15 Questões/GO/Questoes15/Questao7/main.go
package main

//7. Associação Crie uma classe Escola e uma classe Professor. A associação deve permitir que uma escola tenha vários professores, e um professor possa lecionar em várias escolas.
import "fmt"

type Professor struct {
	nome    string
	escolas []*Escola
}

func (p *Professor) AddEscola(e *Escola) {
	p.escolas = append(p.escolas, e)
	e.AddProfessor(p)
}

func (p *Professor) GetEscolas() []*Escola {
	return p.escolas
}

type Escola struct {
	nome        string
	professores []*Professor
}

func (e *Escola) AddProfessor(p *Professor) {
	for _, professor := range e.professores {
		if professor == p {
			return
		}
	}
	e.professores = append(e.professores, p)
}

func (e *Escola) GetProfessores() []*Professor {
	return e.professores
}

func main() {
	prof1 := &Professor{nome: "Kate"}
	prof2 := &Professor{nome: "Jack"}
	prof3 := &Professor{nome: "John"}

	escola1 := &Escola{nome: "Ilha"}
	escola2 := &Escola{nome: "Dharma"}

	prof1.AddEscola(escola1)
	prof2.AddEscola(escola1)
	prof2.AddEscola(escola2)
	prof3.AddEscola(escola2)

	fmt.Println("Escolas da Kate:")
	for _, escola := range prof1.GetEscolas() {
		fmt.Println("   -", escola.nome)
	}

	fmt.Println("Escolas do Jack:")
	for _, escola := range prof2.GetEscolas() {
		fmt.Println("   -", escola.nome)
	}

	fmt.Println("Escolas do John:")
	for _, escola := range prof3.GetEscolas() {
		fmt.Println("   -", escola.nome)
	}

	fmt.Println("Professores da escola Ilha:")
	for _, professor := range escola1.GetProfessores() {
		fmt.Println("   -", professor.nome)
	}

	fmt.Println("Professores da escola Dharma:")
	for _, professor := range escola2.GetProfessores() {
		fmt.Println("   -", professor.nome)
	}
}
